code/greedy_algorithms/src/huffman_coding/huffman_coding.c

```diff
@@ -0,0 +1,211 @@
+#include <stdio.h>
+#include <stdlib.h>
+
+#define MAX_TREE_HT 100
+
+struct MinHeapNode {
+    char data;
+    int freq;
+    struct MinHeapNode *left, *right;
+};
+
+struct MinHeap {
+    int size;
+    int capacity;
+    struct MinHeapNode** array;
+};
+
+struct MinHeapNode* 
+newNode(char data, int freq)
+{
+    struct MinHeapNode* temp = (struct MinHeapNode*)malloc(sizeof(struct MinHeapNode));
+ 
+    temp->left = temp->right = NULL;
+    temp->data = data;
+    temp->freq = freq;
+ 
+    return (temp);
+}
+
+struct MinHeap* 
+createMinHeap(int capacity) 
+{
+    struct MinHeap* minHeap = (struct MinHeap*)malloc(sizeof(struct MinHeap));
+ 
+    minHeap->size = 0;
+    minHeap->capacity = capacity;
+    minHeap->array = (struct MinHeapNode**)malloc(minHeap->capacity * sizeof(struct MinHeapNode*));
+
+    return (minHeap);
+}
+
+void 
+swapMinHeapNode(struct MinHeapNode** a,struct MinHeapNode** b)
+{
+    struct MinHeapNode* t = *a;
+    *a = *b;
+    *b = t;
+}
+
+void 
+minHeapify(struct MinHeap* minHeap, int index)
+{
+    int smallest = index;
+    int left = (2 * index) + 1;
+    int right = (2 * index) + 2;
+ 
+    if (left < minHeap->size && minHeap->array[left]->freq < minHeap->array[smallest]->freq)
+        smallest = left;
+ 
+    if (right < minHeap->size && minHeap->array[right]->freq < minHeap->array[smallest]->freq)
+        smallest = right;
+ 
+    if (smallest != index) {
+        swapMinHeapNode(&minHeap->array[smallest],&minHeap->array[index]);
+        minHeapify(minHeap, smallest);
+    }
+}
+
+int 
+isSizeOne(struct MinHeap* minHeap)
+{
+    return (minHeap->size == 1);
+}
+
+struct MinHeapNode* 
+extractMin(struct MinHeap* minHeap)
+{
+    struct MinHeapNode* temp = minHeap->array[0];
+
+    minHeap->array[0] = minHeap->array[minHeap->size - 1];
+    --minHeap->size;
+    minHeapify(minHeap, 0);
+ 
+    return (temp);
+}
+
+void 
+insertMinHeap(struct MinHeap* minHeap,struct MinHeapNode* minHeapNode)
+{
+    ++minHeap->size;
+    int i = minHeap->size - 1;
+ 
+    while (i && minHeapNode->freq < minHeap->array[(i - 1) / 2]->freq) {
+        minHeap->array[i] = minHeap->array[(i - 1) / 2];
+        i = (i - 1) / 2;
+    }
+ 
+    minHeap->array[i] = minHeapNode;
+}
+
+void
+buildMinHeap(struct MinHeap* minHeap)
+{
+    int n = minHeap->size - 1;
```

Comment:
I would get rid of this variable altogether, and use `minHeap->size - 2` in the loop.